pkg/cloud/externalconn/connection.go

```diff
@@ -22,14 +22,24 @@ import (
 // external resource.
 type Connection interface{}
 
+// ConnectionContext contains the dependencies passed to external connection
+// implementation during creation.
+type ConnectionContext interface {
```

Comment:
does this have to be an interface?
could it be a simple struct?

Reply:
I think it could be a struct but would we just leave some fields unset depending on where we're Dialing from? For example when passing in a `kmsEnv` we don't want to be setting `limiters` that is exclusively an `ExternalStorage` concept.

Reply:
i'm with aditya, i vote interface!

Reply:
> i'm with aditya, i vote interface!

Too bad -- it's a configuratoin struct.  That's what it is.
The fact that some configuration may or may not be applicable is not a reason for additional complexity.

Reply:
> > i'm with aditya, i vote interface!
> 
> Too bad -- it's a configuratoin struct. That's what it is. The fact that some configuration may or may not be applicable is not a reason for additional complexity.

The complexity is intrinsic to the problem of multiple-backends-with-a-singular-API, no? We could manage that complexity through a single struct whose fields are used inconsistently. Or, we could use multiple structs with a single consistently-used interface. Why is the former so much better?

Reply:
fwiw while adding support for a kafka sink I'm leaning towards changing Dial() to take in an `interface{}` because of the CCL/non-CCL boundary and the complications that arise with that. I do expect the signatures to evolve in the next couple of PRs so there will be more chances to tweak it.